### PR TITLE
pwx-38744 | test: add tests for errored snapshot cleanup

### PR DESCRIPTION
**Pull request by strivedi-px**
**What type of PR is this?**
integration-test

**What this PR does / why we need it**:
Closes [PWX-38744](https://purestorage.atlassian.net/browse/PWX-38744). Adds integration tests for the cleanup of snapshots in `Error` state.

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:
No

**Does this change need to be cherry-picked to a release branch?**:
No

**Test run screenshot and pipeline run**
Screenshot from 2024-09-14 15-46-53

Jenkins run: https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%2024.3.1-dev/job/snapshot-csi-k8s-1-25/2614/

### CODE REVIEW

File: test/integration_test/snapshot_test.go [hunk 1]
```diff
@@ -1180,6 +1187,479 @@ func updateRetainValueTest(t *testing.T) {
...
+	Dash.VerifyFatal(t, len(snapStatus), 1, "Should have snapshots for only one policy type")
+	log.InfoD("Validated snapshotschedule %s to have two volumesnapshots", scheduleNameTemp)
+
+	//////////////////////////////////////////////////////////
+	// Find the pods using the PVC to determine the node.
+	// Post determining, stop portworx service on the node.
+	/////////////////////////////////////////////////////////
+	podLabel := make(map[string]string, 0)
+	podLabel["app"] = "mysql"
+	pvcList, err := core.Instance().GetPersistentVolumeClaims(namespace, podLabel)
```
Comment (diptiranjanpx) [inline test/integration_test/snapshot_test.go:1539]: Can you make a common function to stop the driver on a node with input as a PVC?

File: test/integration_test/snapshot_test.go [hunk 2]
```diff
@@ -1180,6 +1187,479 @@ func updateRetainValueTest(t *testing.T) {
...
+	if errorCount == 0 {
+		Dash.Fatal("should have more than one snapshots in Error state")
+	}
+
+	// Wait for 90 seconds more.
+	log.Info("Waiting for 90 seconds more to prevent any transient error")
+	time.Sleep(90 * time.Second)
+
+	// Verify that some of snapshots have been cleaned up.
+	snapshots, err = k8sextops.Instance().ListSnapshots(namespace)
```
Comment (diptiranjanpx) [inline test/integration_test/snapshot_test.go:1615]: Let's add a common function to check in a volumesnapshotschedule , how many are in errored state and also deleted , also can the verify function be common?

File: test/integration_test/snapshot_test.go [hunk 3]
```diff
@@ -1180,6 +1187,479 @@ func updateRetainValueTest(t *testing.T) {
...
+
+	// Deploy a dummy application for taking snapshots.
+	ctx := createApp(t, "delete-error-snapshot-cr-test")
+
+	////////////////////////////////////////////////////
+	// Create schedulepolicy and snapshot schedule.
+	////////////////////////////////////////////////////
+	policyName := "intervalpolicy"
+	scheduleName := "errsnap-schedule"
+	retain := 5
```
Comment (diptiranjanpx) [inline test/integration_test/snapshot_test.go:1212]: Lets make the retain value small like 3, and then lets have in this same test check
1. When first snapshot itself errors out
2. When first 2 snasphots are successful and then the third one errors out
3. When first 3 snaps are successful and the 4th snapshot errors out.

By doing this we can cover allmost all the cases for the schedule functionality wise.

File: test/integration_test/snapshot_test.go [hunk 4]
```diff
@@ -1180,6 +1187,479 @@ func updateRetainValueTest(t *testing.T) {
...
+	// Bring down portworx on the node where the PVC is attached.
+	err = core.Instance().AddLabelOnNode(vol1Node, schedops.PXServiceLabelKey, k8sServiceOperationStop)
+	log.FailOnError(t, err, "Error bringing down node")
+	err = core.Instance().AddLabelOnNode(vol2Node, schedops.PXServiceLabelKey, k8sServiceOperationStop)
+	log.FailOnError(t, err, "Error bringing down node")
+	defer func(node1, node2 string) {
+		// Bring portworx backup up on the node where the PVC is attached.
+		err = core.Instance().AddLabelOnNode(node1, schedops.PXServiceLabelKey, k8sServiceOperationStart)
+		log.FailOnError(t, err, "Error starting PX on node")
+		err = core.Instance().AddLabelOnNode(node2, schedops.PXServiceLabelKey, k8sServiceOperationStop)
```
Comment (strivedi-px) [inline test/integration_test/snapshot_test.go:1584]:
```suggestion
		err = core.Instance().AddLabelOnNode(node2, schedops.PXServiceLabelKey, k8sServiceOperationStart)
```

File: test/integration_test/snapshot_test.go (hunk 2 above)
Comment (strivedi-px) [reply test/integration_test/snapshot_test.go:1498]: Done

File: test/integration_test/snapshot_test.go (hunk 1 above)
Comment (strivedi-px) [reply test/integration_test/snapshot_test.go]: Done

File: test/integration_test/snapshot_test.go (hunk 3 above)
Comment (strivedi-px) [reply test/integration_test/snapshot_test.go:1210]: Can you please elaborate what is meant to be tested here?

Comment (diptiranjanpx) [reply test/integration_test/snapshot_test.go:1210]: I was thinking of handling all the 3 cases with error snapshots in integration test. Anyway we can take it later in another PR.

**Review by diptiranjanpx — APPROVED** [review]

**Pull request merged**